// react-flip-toolkit/lib/Flipper.js
"use strict";

exports.__esModule = true;
exports.PortalContext = exports.FlipContext = undefined;

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _flip = require("./flip");

var _flip2 = _interopRequireDefault(_flip);

var _getFlippedElementPositions = require("./flip/getFlippedElementPositions");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var FlipContext = exports.FlipContext = (0, _react.createContext)("flip");
var PortalContext = exports.PortalContext = (0, _react.createContext)("portal");

var Flipper = function (_Component) {
  _inherits(Flipper, _Component);

  function Flipper(props) {
    _classCallCheck(this, Flipper);

    var _this = _possibleConstructorReturn(this, _Component.call(this, props));

    _this.inProgressAnimations = {};
    _this.flipCallbacks = {};
    return _this;
  }

  Flipper.prototype.getSnapshotBeforeUpdate = function getSnapshotBeforeUpdate(prevProps) {
    if (prevProps.flipKey !== this.props.flipKey) {
      return (0, _getFlippedElementPositions.getFlippedElementPositionsBeforeUpdate)({
        element: this.el,
        // if onExit callbacks exist here, we'll cache the DOM node
        flipCallbacks: this.flipCallbacks,
        inProgressAnimations: this.inProgressAnimations,
        portalKey: this.props.portalKey
      });
    }
    return null;
  };

  Flipper.prototype.componentDidUpdate = function componentDidUpdate(prevProps, prevState, cachedData) {
    if (this.props.flipKey !== prevProps.flipKey) {
      (0, _flip2.default)({
        cachedFlipChildrenPositions: cachedData.flippedElementPositions,
        cachedOrderedFlipIds: cachedData.cachedOrderedFlipIds,
        containerEl: this.el,
        inProgressAnimations: this.inProgressAnimations,
        flipCallbacks: this.flipCallbacks,
        applyTransformOrigin: this.props.applyTransformOrigin,
        spring: this.props.spring,
        debug: this.props.debug,
        portalKey: this.props.portalKey,
        staggerConfig: this.props.staggerConfig,
        handleEnterUpdateDelete: this.props.handleEnterUpdateDelete,
        decisionData: {
          prev: prevProps.decisionData,
          current: this.props.decisionData
        }
      });
    }
  };

  Flipper.prototype.render = function render() {
    var _this2 = this;

    var _props = this.props,
        element = _props.element,
        className = _props.className,
        portalKey = _props.portalKey;

    var Element = element;

    return _react2.default.createElement(
      PortalContext.Provider,
      { value: portalKey },
      _react2.default.createElement(
        FlipContext.Provider,
        { value: this.flipCallbacks },
        _react2.default.createElement(
          Element,
          { className: className, ref: function ref(el) {
              return _this2.el = el;
            } },
          this.props.children
        )
      )
    );
  };

  return Flipper;
}(_react.Component);

Flipper.propTypes = process.env.NODE_ENV !== "production" ? {
  flipKey: _propTypes2.default.any,
  children: _propTypes2.default.node.isRequired,
  spring: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.object]),
  applyTransformOrigin: _propTypes2.default.bool,
  debug: _propTypes2.default.bool,
  element: _propTypes2.default.string,
  className: _propTypes2.default.string,
  portalKey: _propTypes2.default.string,
  staggerConfig: _propTypes2.default.object,
  decisionData: _propTypes2.default.any,
  handleEnterUpdateDelete: _propTypes2.default.func
} : {};

Flipper.defaultProps = {
  applyTransformOrigin: true,
  element: "div"
};

exports.default = Flipper;
